addons/controllers/cpi/cpiconfig_controller.go

```diff
@@ -0,0 +1,314 @@
+	}
+	res, err = r.ReconcileCPIConfigNormal(ctx, req, cpiConfig, cluster)
+	if err != nil {
+		r.Log.Error(err, "Error reconciling CPIConfig to create/patch data values secret")
+	}
+	r.Log.Info("Successfully reconciled CPIConfig")
+	return res, err
+}
+
+// ReconcileCPIConfigNormal triggers when a CPIConfig is not being deleted
+// it ensures the owner reference of the CPIConfig and generates the data values secret for CPI
+func (r *CPIConfigReconciler) ReconcileCPIConfigNormal(ctx context.Context, _ ctrl.Request,
+	cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (ctrl.Result, error) {
+	// add owner reference to CPIConfig if not already added by TanzuClusterBootstrap Controller
+	ownerReference := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+	r.Log.Info("Ensure CPIConfig has the cluster as owner reference")
+	if !clusterapiutil.HasOwnerRef(cpiConfig.OwnerReferences, ownerReference) {
+		r.Log.Info("Adding owner reference to CPIConfig")
+		cpiConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(cpiConfig.OwnerReferences, ownerReference)
+	}
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretName(cluster.Name, constants.CPIAddonName),
+			Namespace: cpiConfig.Namespace,
+		},
+		Type: v1.SecretTypeOpaque,
+	}
+	secret.SetOwnerReferences([]metav1.OwnerReference{ownerReference})
+
+	mutateFn := func() error {
+		secret.Data = make(map[string][]byte)
+		cpiConfigSpec, err := r.mapCPIConfigToDataValues(ctx, cpiConfig, cluster)
+		r.Log.Info("Mutate CPIConfig data values", "data values", cpiConfigSpec)
+		if err != nil {
+			r.Log.Error(err, "Error occurs when mutate CPIConfig data values")
+			return err
+		}
+		yamlBytes, err := yaml.Marshal(cpiConfigSpec)
+		if err != nil {
+			r.Log.Error(err, "Error marshaling CPIConfig to Yaml")
+			return err
+		}
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
+		secret.Data[constants.TKGDataValueFileName] = dataValueBytes
+		return nil
+	}
+	result, err := controllerutil.CreateOrPatch(ctx, r.Client, secret, mutateFn)
+	if err != nil {
+		r.Log.Error(err, "Error creating or patching CPIConfig data values secret")
+		return ctrl.Result{}, err
+	}
+	r.Log.Info(fmt.Sprintf("Resource '%s' data values secret '%s'", constants.CPIAddonName, result))
+	// update the secret reference in CPIConfig status
+	cpiConfig.Status.SecretRef = v1.SecretReference{
+		Name:      secret.Name,
+		Namespace: secret.Namespace,
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ReconcileCPIConfigDelete triggers when a CPIConfig is deleting, currently do nothing
+func (r *CPIConfigReconciler) ReconcileCPIConfigDelete(context.Context, ctrl.Request, *cpiv1alpha1.CPIConfig, *clusterapiv1beta1.Cluster) (res ctrl.Result, retErr error) {
+	return ctrl.Result{}, nil
```

Comment:
Please just return this on the caller and remove this function

Reply:
removed